support Netty io_uring transport

sullis opened the pull request:
- https://en.wikipedia.org/wiki/Io_uring
- https://github.com/netty/netty-incubator-transport-io_uring

sullis commented:
fyi @normanmaurer

normanmaurer commented:
nice...!

sullis commented:
Work in progress. cc: @agoniberg

sullis commented:
Work in progress 
cc: @artgon

sullis commented:
@argha-c  This is a draft PR of io_uring support. WDYT?

argha-c commented on zuul-core/src/main/java/com/netflix/zuul/netty/server/Server.java:
```diff
@@ -378,6 +382,15 @@ private void initializeTransport()
                         workerExecutor,
                         chooserFactory,
                         DefaultSelectStrategyFactory.INSTANCE);
+            } else if (!useNio && ioUringIsAvailable()) {
```
Also, I'd consider adding a toggle if the user wants to explicitly enable this.

sullis replied:
I just added FORCE_IO_URING to Server.java. Let me know what you think.

argha-c commented on zuul-io-uring-test/src/test/java/com/netflix/zuul/netty/server/IoUringTest.java:
```diff
@@ -0,0 +1,141 @@
+        assertEquals(2, ioUringChannelCount.get());
+    }
+
+    private static void openHttpConnection(final int port) {
```
This doesn't actually validate a proper request-response?

sullis replied:
I replaced `openHttpConnection` with `checkConnection`

artgon commented on zuul-core/src/main/java/com/netflix/zuul/netty/server/Server.java:
```diff
@@ -354,8 +365,19 @@ private void initializeTransport()
-            if (!useNio && epollIsAvailable()) {
+            final boolean useNio = FORCE_NIO.get();
+            final boolean useIoUring = FORCE_IO_URING.get();
+            if (useIoUring) {
```
Shouldn't this also check if it's available?

sullis replied:
Fixed.

artgon commented on zuul-io-uring-test/src/test/java/com/netflix/zuul/netty/server/IoUringTest.java:
```diff
@@ -0,0 +1,159 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(IoUringTest.class);
+
+    @Test
+    public void testIoUringServer() throws Exception {
```
What's the goal of this test? Just verifying that the server starts up?

sullis replied:
I just added a JavaDoc comment that explains the goal of the test.

sullis commented:
I rebased this PR with the master branch.

Ready for review @argha-c

artgon commented:
I think this is really cool, but seems like it's "incubator" mode so we're unlikely to run it in production (yet). I'm open to merging it to allow people to play around with it but it won't have the kind of verification that the rest of the code has.

sullis commented:
> I think this is really cool, but seems like it's "incubator" mode so we're unlikely to run it in production (yet). I'm open to merging it to allow people to play around with it but it won't have the kind of verification that the rest of the code has.

Agreed.  It is an "incubator" feature.  I think it would be good to merge it so that folks can play around with it.

sullis commented:
I resolved all merge conflicts.

argha-c commented:
@sullis : Thanks for addressing the feedback so far. We're definitely open to merging it, and appreciate your contributions.

If you could consolidate and exercise the tests in our regular module (now that we have the FP), I think it'd be more valuable than introducing a separate test module altogether.

sullis commented:
> @sullis : Thanks for addressing the feedback so far. We're definitely open to merging it, and appreciate your contributions.
> 
> If you could consolidate and exercise the tests in our regular module (now that we have the FP), I think it'd be more valuable than introducing a separate test module altogether.

I moved IoUringTest.java to zuul-core

Is this ok or should I move the test case into ServerTest.java ?

argha-c commented:
LGTM. Please rebase and resolve conflicts.

sullis commented:
> LGTM. Please rebase and resolve conflicts.

Done.
